storage/store.go

```diff
@@ -65,17 +82,321 @@ type Store interface {
 	) error
 
 	// CollectionByID gets the collection (transaction IDs only) with the given ID.
-	CollectionByID(flowgo.Identifier) (flowgo.LightCollection, error)
+	CollectionByID(ctx context.Context, collectionID flowgo.Identifier) (flowgo.LightCollection, error)
 
 	// TransactionByID gets the transaction with the given ID.
-	TransactionByID(flowgo.Identifier) (flowgo.TransactionBody, error)
+	TransactionByID(ctx context.Context, transactionID flowgo.Identifier) (flowgo.TransactionBody, error)
 
 	// TransactionResultByID gets the transaction result with the given ID.
-	TransactionResultByID(flowgo.Identifier) (types.StorableTransactionResult, error)
+	TransactionResultByID(ctx context.Context, transactionID flowgo.Identifier) (types.StorableTransactionResult, error)
 
 	// LedgerViewByHeight returns a view into the ledger state at a given block.
-	LedgerViewByHeight(blockHeight uint64) *delta.View
+	LedgerViewByHeight(ctx context.Context, blockHeight uint64) *delta.View
 
 	// EventsByHeight returns the events in the block at the given height, optionally filtered by type.
-	EventsByHeight(blockHeight uint64, eventType string) ([]flowgo.Event, error)
+	EventsByHeight(ctx context.Context, blockHeight uint64, eventType string) ([]flowgo.Event, error)
+}
+
+type StorageBackendKeys interface {
+	StorageKey(key string) string
+	LatestBlockKey() []byte
+	BlockHeightKey(height uint64) []byte
+	IdentifierKey(id flowgo.Identifier) []byte
+	EventKey(blockHeight uint64, txIndex, eventIndex uint32, eventType flowgo.EventType) []byte
+}
+
+type StorageBackendData interface {
+	GetBytes(ctx context.Context, store string, key []byte) ([]byte, error)
+	SetBytes(ctx context.Context, store string, key []byte, value []byte) error
+	SetBytesWithVersion(ctx context.Context, store string, key []byte, value []byte, version uint64) error
+	GetBytesAtVersion(ctx context.Context, store string, key []byte, version uint64) ([]byte, error)
+}
+
+type StorageBackendKeysImpl struct {
+}
+
+func (s *StorageBackendKeysImpl) StorageKey(key string) string {
+	return key
+}
+
+func (s *StorageBackendKeysImpl) LatestBlockKey() []byte {
+	return []byte("latest_block_height")
+}
+
+func (s *StorageBackendKeysImpl) BlockHeightKey(blockHeight uint64) []byte {
+	return []byte(fmt.Sprintf("%032d", blockHeight))
+}
+
+func (s *StorageBackendKeysImpl) IdentifierKey(id flowgo.Identifier) []byte {
+	return []byte(fmt.Sprintf("%x", id))
+}
+
+func (s *StorageBackendKeysImpl) EventKey(blockHeight uint64, txIndex, eventIndex uint32, eventType flowgo.EventType) []byte {
+	return []byte(fmt.Sprintf(
+		"%032d-%032d-%032d-%s",
+		blockHeight,
+		txIndex,
+		eventIndex,
+		eventType,
+	))
+}
+
+type StoreImpl struct {
```

Comment:
Avoid exporting and rename if possible to drop `Impl`, I suggest maybe `defaultStore` or anything like that

Reply:
should I flatten redis , badger stores and move to storage level ?

Reply:
I don't mind them being there, I think it shows nice isolation of each implementation.

Reply:
Yeah I mean in the context of exporting, I was exporting for then to be able to use, but my go knowledge is limited, is there another way?

Reply:
But where do you use it? I didn't find any usage. If you think it needs to be used then leave it exported but I only saw usage of the Store interface.